tests/test_reload.py

```diff
@@ -54,7 +54,6 @@ def test_reload_inc_device_no_exists(self):
 
     def test_reload_unverified_snapshot(self):
         util.unmount(self.mount)
-        self.addCleanup(util.mount, self.device, self.mount)
```

Comment:
These tests are modified copy-paste from other tests. 
For instance this test `test_reload_unverified_snapshot` is very close to `test_destroy_unverified_snapshot`. And `self.addCleanup(util.mount, self.device, self.mount)` is used in `test_setup.py`, `test_destroy.py` and here in `test_reload.py`.
Why do we need to remove this cleanup just in in `test_reload.py`? And how `umount` operation is handled now?

Reply:
Then I'm not sure it's correctly implemented in other tests. In this one we manually mount and dismount the volume. After the test it will be in the initial state (mounted). If it, f.e., fails to unmount during the test, then we failed it anyway. In my opinion, there's no need in this additional umount/mount operation, let's just do it straight.

Reply:
Well, I just did a small investigation and here it is.
### Background
The tests are doing umount at the beginning and there is a cleanup to mount the test device back in any case. https://github.com/elastio/elastio-snap/blob/e4c0e7d15423c57fed39db9e60aa425ebbc47e70/tests/test_reload.py#L55-L57
Why it's necessary? Because of tear down:
https://github.com/elastio/elastio-snap/blob/e4c0e7d15423c57fed39db9e60aa425ebbc47e70/tests/devicetestcase.py#L61-L62

### Experiment
Let's say `test_reload_unverified_snapshot` (picked up randomly) has hailed on line 59..63 before `mount` on line 66.
https://github.com/elastio/elastio-snap/blob/e4c0e7d15423c57fed39db9e60aa425ebbc47e70/tests/test_reload.py#L59-L66
I've added `self.fail("experimental failure")` after the line `self.check_snap_info(0, elastio_snap.State.UNVERIFIED, True)`.
And there is an execution of `test_reload` set on `master` branch with the cleanups by `sudo ./elio-test.sh -t test_reload`:
```
elastio@pop-os:~/elastio-snap/tests$ sudo ./elio-test.sh -t test_reload

elastio-snap: 7612806
kernel: 6.0.12-76060006-generic
filesystem: ext4
gcc: 11.3.0-1ubuntu1~22.04)
bash: 5.1.16(1)-release
python: Python 3.10.6

test_reload_inc_device_no_exists (test_reload.TestReload) ... ok
test_reload_inc_invalid_minor (test_reload.TestReload) ... ok
test_reload_inc_volume_path_is_dir (test_reload.TestReload) ... ok
test_reload_snap_device_no_exists (test_reload.TestReload) ... ok
test_reload_snap_invalid_minor (test_reload.TestReload) ... ok
test_reload_snap_volume_path_is_dir (test_reload.TestReload) ... ok
test_reload_unverified_incremental (test_reload.TestReload) ... FAIL
test_reload_unverified_snapshot (test_reload.TestReload) ... ok
test_reload_verified_inc (test_reload.TestReload) ... ok
test_reload_verified_snapshot (test_reload.TestReload) ... ok

======================================================================
FAIL: test_reload_unverified_incremental (test_reload.TestReload)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_reload.py", line 81, in test_reload_unverified_incremental
    self.fail("experimental failure");
AssertionError: experimental failure

----------------------------------------------------------------------
Ran 10 tests in 3.683s

FAILED (failures=1)

```
As you can see, test `test_reload_unverified_incremental` has failed. And all tests after it are passed.

Then I did the same steps on your branch with the removed cleanups:
```
elastio@pop-os:~/elastio-snap/tests$ sudo ./elio-test.sh -t test_reload

elastio-snap: b8115ff
kernel: 6.0.12-76060006-generic
filesystem: ext4
gcc: 11.3.0-1ubuntu1~22.04)
bash: 5.1.16(1)-release
python: Python 3.10.6

test_reload_inc_device_no_exists (test_reload.TestReload) ... ok
test_reload_inc_invalid_minor (test_reload.TestReload) ... ok
test_reload_inc_volume_path_is_dir (test_reload.TestReload) ... ok
test_reload_snap_device_no_exists (test_reload.TestReload) ... ok
test_reload_snap_invalid_minor (test_reload.TestReload) ... ok
test_reload_snap_volume_path_is_dir (test_reload.TestReload) ... ok
test_reload_unverified_incremental (test_reload.TestReload) ... FAIL
test_reload_unverified_snapshot (test_reload.TestReload) ... Command umount /tmp/elastio-snap_004 has failed (32): umount: /tmp/elastio-snap_004: not mounted.

ERROR
test_reload_verified_inc (test_reload.TestReload) ... FAIL
test_reload_verified_snapshot (test_reload.TestReload) ... FAIL
Command umount /tmp/elastio-snap_004 has failed (32): umount: /tmp/elastio-snap_004: not mounted.

ERROR

======================================================================
ERROR: test_reload_unverified_snapshot (test_reload.TestReload)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_reload.py", line 56, in test_reload_unverified_snapshot
    util.unmount(self.mount)
  File "/home/elastio/elastio-snap/tests/util.py", line 39, in unmount
    raise subprocess.CalledProcessError(p.returncode, cmd, "Command failed")
subprocess.CalledProcessError: Command '['umount', '/tmp/elastio-snap_004']' returned non-zero exit status 32.

======================================================================
ERROR: tearDownClass (test_reload.TestReload)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/devicetestcase.py", line 62, in tearDownClass
    util.unmount(cls.mount)
  File "/home/elastio/elastio-snap/tests/util.py", line 39, in unmount
    raise subprocess.CalledProcessError(p.returncode, cmd, "Command failed")
subprocess.CalledProcessError: Command '['umount', '/tmp/elastio-snap_004']' returned non-zero exit status 32.

======================================================================
FAIL: test_reload_unverified_incremental (test_reload.TestReload)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_reload.py", line 78, in test_reload_unverified_incremental
    self.fail("experimental failure");
AssertionError: experimental failure

======================================================================
FAIL: test_reload_verified_inc (test_reload.TestReload)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_reload.py", line 111, in test_reload_verified_inc
    self.assertEqual(elastio_snap.setup(self.minor, self.device, self.cow_full_path, ignore_snap_errors=True), 0)
AssertionError: 22 != 0

======================================================================
FAIL: test_reload_verified_snapshot (test_reload.TestReload)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_reload.py", line 87, in test_reload_verified_snapshot
    self.assertEqual(elastio_snap.setup(self.minor, self.device, self.cow_full_path, ignore_snap_errors=True), 0)
AssertionError: 22 != 0

----------------------------------------------------------------------
Ran 10 tests in 1.251s

FAILED (failures=3, errors=2)

```
All tests after the `test_reload_unverified_incremental` are failed too due to `umout` failure in the `tearDownClass` method.

Obviously, the 1st kind of behavior is better. 

### Conclusion
The current changes with the removal of cleanups ` self.addCleanup(util.mount, self.device, self.mount)` are not correct.
You have to bring them back or modify `tearDownClass` method with the check if this mount exists before `umount` call.
 https://github.com/elastio/elastio-snap/blob/e4c0e7d15423c57fed39db9e60aa425ebbc47e70/tests/devicetestcase.py#L61-L62

Reply:
@e-kov thanks for the investigation, it was very helpful. I think I will choose way #1 (bring them back) and just leave them as they were.

Reply:
@e-kov  thanks for the investigation, it was very helpful. I have chosen way #1 (to bring them back). These cleanups make all the difference though, as the issue is reproduced more often with them. This, according to my investigation, is related to the cleanup sequence and delayed LVM operations. I've made a small change (which is crucial in this case): to make sure that `destroy` is called at the very end, after the umount/mount sequence is complete. Now the tests are stable.

Reply:
Great! Thank you.